### PR TITLE
Plugins for Uoccin tracking app

### PR DESCRIPTION
**Pull request by tarzasai**
These plugins cover all the tracking functionality of third party services like trakt.tv. All about our watchlist, collections, watched stuff and so on will be stored in a single file, ```uoccin.json```, in this structure:
```
{
    "movies": {
        "tt2713180": {
            "name": "Fury",
            "watchlist": true,
            "collected": false,
            "watched": false,
            "rating": 0,
            "tags": [ "IMDB" ]
        },
        "tt2278388": {
            "name": "The Grand Budapest Hotel",
            "watchlist": false,
            "collected": true,
            "watched": false,
            "rating": 0,
            "subtitles": [ "eng", "ita" ]
        }
    },
    "series": {
        "265766": {
            "name": "Penny Dreadful",
            "watchlist": true,
            "collected": {},
            "watched": {
                "1": [ 1, 2, 3, 4, 5, 6, 7, 8 ],
                "2": [ 1 ]
            },
            "rating": 4,
            "tags": [ "favorite", "hires" ]
        },
        "279121": {
            "name": "The Flash (2014)",
            "watchlist": true,
            "collected": {
                "1": {
                    "17": [ "eng", "ita" ],
                    "18": [ "eng", "ita" ],
                    "19": [ "eng" ],
                    "20": [ "eng", "ita" ],
                    "21": []
                }
            },
            "watched": {
                "1": [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 ]
            },
            "rating": 1,
            "tags": [ "favorite" ]
        }
    }
}
```
There are 2 plugins reading this file:

**uoccin_lookup** is a metainfo plugin to recover all the info about movies, series and episodes:
```
  ut1:
    imdb_lookup: yes
    uoccin_lookup: '{{ secrets.uoccin.path }}'
    mock:
      - { title: 'fury' }

>>> output:
  title            : fury
  url              : http://localhost/mock/1869571976
  original_url     : http://localhost/mock/1869571976
  imdb_id          : tt2713180
  imdb_name        : Fury
  [...more info from imdb_lookup...]
  quality          : unknown
  task             : ut1
  uoccin_collected : False
  uoccin_rating    : 0
  uoccin_subtitles : []
  uoccin_tags      : [IMDB]
  uoccin_watched   : False
  uoccin_watchlist : True
```
and here's about episodes:
```
  ut2:
    metainfo_series: yes
    thetvdb_lookup: yes
    imdb_lookup: yes
    uoccin_lookup: '{{ secrets.uoccin.path }}'
    mock:
      - { title: 'penny dreadful s01e03' }

>>> output:
  title            : penny dreadful s01e03
  url              : http://localhost/mock/1133637368
  original_url     : http://localhost/mock/1133637368
  series_episode   : 3
  series_episodes  : 1
  series_guessed   : True
  series_id        : S01E03
  series_id_type   : ep
  series_name      : Penny Dreadful
  [...lot of stuff from tvdb_lookup...]
  uoccin_collected : False
  uoccin_rating    : 4
  uoccin_subtitles : []
  uoccin_tags      : [favorite]
  uoccin_watched   : True
  uoccin_watchlist : True
  zap2it_id        : SH01680293
```

Then there is **uoccin_emit**, the input plugin to use with configure_series:
```
  get_series:
    metainfo_series: yes
    thetvdb_lookup: yes
    uoccin_lookup: '{{ secrets.uoccin.path }}'
    sequence:
      - configure_series:
          settings:
            quality: 720p hdtv+
          from:
            uoccin_emit:
              path: '{{ secrets.uoccin.path }}'
              type: series
              tags: [ 'hires' ]
      - configure_series:
          settings:
            quality: <720p hdtv+
          from:
            uoccin_emit:
              path: '{{ secrets.uoccin.path }}'
              type: series
              tags: [ 'hires', 'netflix' ]
              check_tags: none
      - configure_series:
          settings:
            quality: webrip
          from:
            uoccin_emit:
              path: '{{ secrets.uoccin.path }}'
              type: series
              tags: [ 'netflix' ]
```
(it can emit the movies watchlist too)

Now to *write* in uoccin.json we have 6 more plugins:
- **uoccin_watchlist_add**
- **uoccin_watchlist_remove**
- **uoccin_collection_add**
- **uoccin_collection_remove**
- **uoccin_watched_true**
- **uoccin_watched_false**

They do just what their names tell they do. In addition, for each task they write all the modifications they've made in a little file that the Android app Uoccin (*) will use to stay synchronized. It's something like this:
```
1431093328971|movie|tt346578|watchlist|true
1431093328971|movie|tt283759|collected|false
1431093329029|series|80379|watchlist|true
1431093329033|series|80379|tags|pippo,pluto
1431175098984|series|80379.8.24|watched|true
1431198108547|series|272135.2.5|collected|true
1431198108565|series|272135.2.5|subtitles|eng,ita
```

The last plugin, **uoccin_reader**, is to process the files coming from the Uoccin app - or another flexget machine or whatever software crazy enough to use this "protocol". It requires a task like this:
```
  uoccin_sync:
    seen: local
    find:
      path:
        - '{{ secrets.uoccin.path }}\device.{{ secrets.uoccin.uuid }}'
      regexp: '.*\.diff$'
    accept_all: yes
    uoccin_reader:
      uuid: '{{ secrets.uoccin.uuid }}'
      path: '{{ secrets.uoccin.path }}'
```

(*) about the app: it's open source (https://github.com/tarzasai/Uoccin) and it'll be free and ads-free. I'm still working on it, but all the main features (searching, editing, synchronization via Google Drive) already work. ~~Alpha and beta releases are available joining the google+ community https://plus.google.com/u/0/communities/111036056585214228136 (some screenshots available there)~~ ...and of course I would be very happy if someone would like to help the development.

### CODE REVIEW

**Comment by tarzasai** [comment]
just wanted to say the app is now out of beta, free at https://play.google.com/store/apps/details?id=net.ggelardi.uoccin :)

File: flexget/plugins/services/uoccin_plugins.py [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+from flexget import plugin
+from flexget.entry import Entry
+from flexget.event import event
+from flexget.utils import json
+
+try:
+    from flexget.plugins.api_tvdb import lookup_series
+except ImportError:
+    raise plugin.DependencyError(issued_by='myepisodes', missing='api_tvdb',
+                                 message='myepisodes requires the `api_tvdb` plugin')
```
Comment (gazpachoking) [inline flexget/plugins/services/uoccin_plugins.py]: myepisodes? :wink:

**Comment by gazpachoking** [comment]
I'm thinking the metainfo plugin should register lazy fields rather than doing the direct lookup.

**Comment by tarzasai** [comment]
Yes, it's just a big dictionary with tvdb_id/imdb_id as keys.

**Comment by liiight** [comment]
@tarzasai I think it'll be preferable to split the plugins to match our current directory phase structure. We're planning an overhaul of directory structure (one day...)  and it would be better to maintain consistency.

**Comment by tarzasai** [comment]
well, I can move the emit plugin (input folder) and the lookup (metainfo), but all the other plugins descend from (or use) the process class, so they can stay in the output folder but in a single file.
about the tests, I've never did one but I'll try.

**Comment by liiight** [comment]
Tests would be great. If you need help or have any question I'm sure you'll know how to contact me

**Comment by tarzasai** [comment]
I don't know what the hell happened with that merge-remote-tracking, can I revert it?

**Comment by gazpachoking** [comment]
@tarzasai That commit seems fine, looks like you just merged in the latest changes from develop in to your branch.

**Comment by tarzasai** [comment]
uhm yes but I didn't expect to see it here... I know very few things about git. I'll try to add some tests so.

**Comment by cvium** [comment]
If you want to avoid merging 'master' or 'develop' into your PR in the future, you'll need to do a rebase (which everyone should be doing anyway).

**Comment by tarzasai** [comment]
about the tests: I'm looking at some inside the tests folder, how to execute them? these plugins read and write an external file, is it a problem?

**Comment by liiight** [comment]
Sorry for not responding. 
Nose tests [doc](https://nose.readthedocs.org/en/latest/).
External files are not a problem, just commit them as usual.

**Comment by tarzasai** [comment]
I've installed nose and run `nosetests` inside the Fleget folder, got a `ERROR: Failure: ImportError (No module named mock)`, in Flexget\tests\__init__.py.
BTW can you tell me if this looks right? http://pastebin.com/ANK1mxjJ

**Comment by liiight** [comment]
You need to install development dependencies in your virtual env:
`pip install -r dev-requirements.txt`

**Comment by tarzasai** [comment]
> This kinda broad IMO

`uoccin_watchlist` is just a boolean field :)

> that won't have a list to remove from when it start.

I see, so I need a ready sample file in the tests folder.

**Comment by tarzasai** [comment]
uhm, I was using `ts = int((datetime.utcnow() - datetime(1970, 1, 1)).total_seconds() * 1000)` to get the long value as java's `System.currentTimeMillis()` (i.e. 1456698635487).
Is there a better way to do it - usable in python 2.6?

**Comment by liiight** [comment]
I got no problem merging this if you feel it's ready

**Comment by tarzasai** [comment]
yup, it is ready.

**Comment by liiight** [comment]
cool. two things now that it's merged: :smile: 
- You should take a look into our [list interface PR](https://github.com/Flexget/Flexget/pull/854). This could definitely fit that schema IMO.
- You can add an API endpoint for this, and then your client app could directly query flexget and get that data. 

If you need help or more info on any of those, i'm sure you know where to find me :stuck_out_tongue_winking_eye:

**Comment by tarzasai** [comment]
yes, I'm definitely interested in those things.
PS: I'll add some wiki pages for now, do I have the permission to link the app store page?

**Comment by liiight** [comment]
to our wiki? sure, you don't need permission for that

**Comment by tarzasai** [comment]
**from** the wiki to the app page; sadly I can't link flexget from google play, the app would be banned in a matter of seconds :P

**Comment by liiight** [comment]
oh. I don't really see a problem with that. and it's always better to ask for forgiveness than permission anyway...

**Pull request merged**